File: 79_WordSearch.java
class WordSearch {
  public static void main(String[] args) {
    WordSearch obj = new WordSearch();
    char[][] board = new char[][] {
      {'A','B','C','E'},
      {'S','F','C','S'},
      {'A','D','E','E'}
    };

    System.out.println(obj.exist(board, "ABCCED"));
    System.out.println(obj.exist(board, "SEE"));
    System.out.println(obj.exist(board, "ABCB"));
  }
  int m, n;
  int[] dir_row, dir_clm;
  boolean[][] seen;
  public boolean exist(char[][] board, String word) {
    m = board.length;
    n = m==0? 0 : board[0].length;
    seen = new boolean[m][n];

    dir_row = new int[] {1,-1,0,0};
    dir_clm = new int[] {0,0,1,-1};

    for (int i=0; i<m; ++i)
      for (int j=0; j<n; ++j)
        if (dfs(i, j, 0, board, word))
          return true;

    return false;
  }

  boolean dfs(int i, int j, int idx, char[][] board, String word) {
    if (idx==word.length())
      return true;

    if (i<0 || i>=m || j<0 || j>=n || seen[i][j] || board[i][j]!=word.charAt(idx))
      return false;

    seen[i][j] = true;

    boolean flag = false;
    for (int k=0; k<4; ++k) {
      flag = dfs(i+dir_row[k], j+dir_clm[k], idx+1, board, word);
      if (flag) break;
    }

    seen[i][j] = false;
    return flag;
  }
}

File: 79_WordSearch.cpp
#include "utils3.hpp"

static const vector<vector<int>> dirs {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};

class Solution {
  private:
    int m, n;
    bool flag;

    void dfs(const char* w, int i, int j, vector<vector<char>>& board, vector<vector<bool>>& visited) {
      if (*w == '\0') {
        flag = true;
        return;
      }
      else if (flag || i < 0 || i == m || j < 0 || j == n || visited[i][j] || board[i][j] != *w)
        return;


      ++w;
      visited[i][j] = true;

      for (auto& d : dirs) 
        dfs(w, i + d[0], j + d[1], board, visited);

      visited[i][j] = false;
    }

  public:
    bool exist(vector<vector<char>>& board, string word) {
      m = board.size();
      n = board.empty() ? 0 : board.front().size();
      flag = false;

      vector<vector<bool>> visited(m, vector<bool>(n));

      for (int i = 0; i < m; ++i)
        for (int j = 0; j < n; ++j)
          dfs(word.c_str(), i, j, board, visited);

      return flag;
    }
};


int main() {
  UFUNCS(Solution::exist);
  return 0;
}